File: frontend/src/services/categories.ts
import http from "../apis/http";
import type { ListParams, Page } from "../types/pagination";
import type {
  Category,
  CategoryCreate,
  CategoryUpdate,
} from "../types/domain";

export type CategoryBatchAction = "enable" | "disable" | "delete";

export interface CategoryBatchItem {
  id: number;
}

export async function listCategories(params: ListParams) {
  const { data } = await http.get<Page<Category>>("/categories", { params });
  return data;
}

export async function listCategoryTree() {
  const { data } = await http.get<Category[]>("/categories/tree");
  return data;
}

export async function createCategory(payload: CategoryCreate) {
  const { data } = await http.post<Category>("/categories", payload);
  return data;
}

export async function updateCategory(categoryId: number, payload: CategoryUpdate) {
  const { data } = await http.put<Category>(`/categories/${categoryId}`, payload);
  return data;
}

export async function deleteCategory(categoryId: number) {
  await http.delete(`/categories/${categoryId}`);
}

export async function importCategories(file: File) {
  const formData = new FormData();
  formData.append("file", file);
  const { data } = await http.post("/categories/import", formData);
  return data as {
    created: number;
    updated: number;
    skipped: number;
    errors?: { row: number; detail: string }[];
  };
}

export async function exportCategories() {
  const response = await http.get<Blob>("/categories/export", {
    responseType: "blob",
  });
  return response.data;
}

export async function batchCategories(
  action: CategoryBatchAction,
  items: CategoryBatchItem[],
) {
  const { data } = await http.post("/categories/batch", { action, items });
  return data as {
    updated: number;
    deleted: number;
    skipped: number;
    errors?: { id: number; detail: string }[];
  };
}

File: backend/app/apis/v1/endpoints/categories.py
from datetime import datetime

from fastapi import APIRouter, Depends, File, UploadFile, status
from fastapi.responses import StreamingResponse
from sqlalchemy.orm import Session

from app.apis.deps import get_current_user, get_db, require_scopes
from app.models.user import User
from app.schemas.category import (
    CategoryBatchAction,
    CategoryBatchResult,
    CategoryCreate,
    CategoryOut,
    CategoryTreeOut,
    CategoryUpdate,
)
from app.schemas.pagination import Page, PageParams
from app.services import categories as category_service

router = APIRouter()


@router.get(
    "",
    dependencies=[Depends(require_scopes(["category:read"]))],
    response_model=Page[CategoryOut],
)
def list_categories(
    params: PageParams = Depends(),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    items, total = category_service.list_categories(db, params)
    return Page(items=items, total=total, page=params.page, page_size=params.page_size)


@router.get(
    "/tree",
    dependencies=[Depends(require_scopes(["category:read"]))],
    response_model=list[CategoryTreeOut],
)
def list_category_tree(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return category_service.list_category_tree(db)


@router.post(
    "",
    dependencies=[Depends(require_scopes(["category:write"]))],
    response_model=CategoryOut,
    status_code=status.HTTP_201_CREATED,
)
def create_category(
    payload: CategoryCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return category_service.create_category(db, payload)


@router.post(
    "/import",
    dependencies=[Depends(require_scopes(["category:write"]))],
)
def import_categories(
    file: UploadFile = File(...),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return category_service.import_categories(db, file)


@router.get(
    "/export",
    dependencies=[Depends(require_scopes(["category:read"]))],
)
def export_categories(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    output = category_service.export_categories(db)
    filename = f"categories_{datetime.utcnow().date().isoformat()}.xlsx"
    headers = {"Content-Disposition": f"attachment; filename={filename}"}
    return StreamingResponse(
        output,
        media_type="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        headers=headers,
    )


@router.post(
    "/batch",
    dependencies=[Depends(require_scopes(["category:write"]))],
    response_model=CategoryBatchResult,
)
def batch_categories(
    payload: CategoryBatchAction,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return category_service.batch_categories(db, payload)


@router.put(
    "/{category_id}",
    dependencies=[Depends(require_scopes(["category:write"]))],
    response_model=CategoryOut,
)
def update_category(
    category_id: int,
    payload: CategoryUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return category_service.update_category(db, category_id, payload)


@router.delete(
    "/{category_id}",
    dependencies=[Depends(require_scopes(["category:write"]))],
    status_code=status.HTTP_204_NO_CONTENT,
)
def delete_category(
    category_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    category_service.delete_category(db, category_id)
    return None
